lib/puppet/application/filebucket.rb

```diff
@@ -148,6 +158,29 @@ def restore
     @client.restore(file, md5)
   end
 
+  def diff
+    checksum_a = args.shift
+    checksum_b = args.shift
+    file_a = nil
+    file_b = nil
+    if checksum_a && Puppet::FileSystem.exist?(checksum_a)
```

Comment:
Rather than `Dipper#diff` taking four arguments and having a somewhat complicated implementation, I think it should take only two (e.g. left/right) and these checks for file existence should be moved there.  If either does not exist, download the file to a temp file and use the temp file as the argument to `Puppet::Util::Diff.diff` (i.e. only one call).